gitpod-io/gitpod
# [ws-scheduler] Don't schedule to nodes that don't have basic services running

Gitpod requires several services to run on a workspace node. This PR ensures that we only schedule to nodes that have these services running.

### How it works
We introduce an `gitpod.io/requiredNodeServices` annotation on workspace pods. ws-scheduler uses this annotation to determine which services (i.e. pods) need to run on a node for that node to be viable. If a pod has a `gitpod.io/nodeService` label, ws-scheduler registers it as service on its node. See `TestRequiredServices` for more details.

With this new mechanism we replace part of the node-daemon node patching. The remainder will go away once everything runs on registry facade.

### Why not PodAffinity?
What we've built here would normally be implemented with PodAffinity. However, PodAffinity is complex and difficult to implement in the scheduler. We might add PodAffinity support in the future to ws-scheduler, at which point we'd remove this mechanism.

### How to test
1. Start a workspace
2. Make ws-daemon unavilable on a some node and start a new workspace - it should be scheduled to a different node
3. Make registry-facade unavailable on some node and start a new workspace with feature preview - it should start on a new node

## ToDo
Supervisor port unit tests fail - this prevents the branch from building
